### PR TITLE
[watsonstt] initial contribution

### PR DESCRIPTION
**Pull request by GiviMAD**
This add-on allows to use IBM Watson Speech-to-Text service within openHAB.
I have tested it using changes from the @lolodomo's [PR](https://github.com/openhab/openhab-core/pull/2693).
Hope everything is in place.

### CODE REVIEW

**Comment by lolodomo** [comment]
Going to the third party link, Zi can read that: :

> On 9 August 2021, IBM announced the deprecation of the Natural Language Classifier service.The service will no longer be available from 8 August 2022. As of 9 September 2021, you will not be able to create new instances. Existing instances will be supported until 8 August 2022. Any instance that still exists on that date will be deleted.

Does it really worth adding something that will die soon?

**Comment by GiviMAD** [comment]
> Does it really worth adding something that will die soon?
I don't think that the " Natural Language Classifier" service is the same that the "Speech to Text" service.
https://cloud.ibm.com/docs/speech-to-text can't find anything here about it been disposed.

**Comment by GiviMAD** [comment]
Free tier for this service it quite better than the google one. That's my main motivation for using it.

**Comment by lolodomo** [comment]
The comment about deprecation iconcerns "Natural Language Classifier". Maybe this is not something used by the Speech to Text ?

File: bundles/org.openhab.voice.watsonstt/NOTICE [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+== Source Code
+
+https://github.com/openhab/openhab-addons
+
+== Third-party Content
+ 
+com.ibm.watson: speech-to-text
+* License: Apache 2.0 License
+* Project: https://github.com/watson-developer-cloud/java-sdk
+* Source: https://github.com/watson-developer-cloud/java-sdk/tree/master/speech-to-text
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/NOTICE:20]: A final return to new line is probably missing.

File: bundles/org.openhab.voice.watsonstt/README.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
+# IBM Watson Speech-to-Text
+
+Watson STT Service uses the non-free IBM Watson Speech-to-Text API to transcript audio data to text. 
+Be aware, that using this service may incur cost on your IBM account.
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/README.md:4]: Be aware that ...

File: bundles/org.openhab.voice.watsonstt/README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+You can find pricing information on [its page](https://www.ibm.com/cloud/watson-speech-to-text/pricing).
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/README.md:5]: this page

**Comment by GiviMAD** [comment]
> The comment about deprecation iconcerns "Natural Language Classifier". Maybe this is not something used by the Speech to Text ?

I think is a service that you can use to analyze the transcriptions, but I think the speech to text service is not directly affected by its deprecation.

File: bundles/org.openhab.voice.watsonstt/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+## Obtaining Credentials
+
+Before you can use this add-on, you should create a Speech-to-Text instance in the IBM Cloud service. 
+
+* Go to the following [link](https://cloud.ibm.com/catalog/services/speech-to-text) and create the instance in your desired region.
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/README.md:11]: Are we talking about instances that will be deleted in 2022 ?

File: bundles/org.openhab.voice.watsonstt/README.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+* After the instance is created you should be able to view its url and api key.
+
+## Configuration
+
+### Authentication Configuration
+
+Using your favorite configuration UI to edit **Settings / Other Services - IBM Watson Speech-to-Text** and set:
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/README.md:18]: Use your ...

File: bundles/org.openhab.voice.watsonstt/README.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+* **Api Key** - Api key for Speech-to-Text instance created on IBM Cloud.
+* **Instance Url** - Url for Speech-to-Text instance created on IBM Cloud.
+
+### Speech to Text Configuration
+
+Using your favorite configuration UI to edit **Settings / Other Services - IBM Watson Speech-to-Text**:
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/README.md:25]: Use your ...

File: bundles/org.openhab.voice.watsonstt/src/main/resources/OH-INF/config/config.xml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+		<parameter name="speechDetectorSensitivity" type="decimal" min="0" max="1" step="0.1" groupName="stt">
+			<label>Speech Detector Sensitivity</label>
+			<description>Use the parameter to suppress word insertions from music, coughing, and other non-speech events.</description>
+			<default>0.5</default>
+		</parameter>
+		<parameter name="inactivityTimeout" type="integer" unit="s" groupName="stt">
+			<label>Inactivity Timeout</label>
+			<description>The time in seconds after which, if only silence (no speech) is detected in the audio, the connection is
+				closed.</description>
+			<unitLabel>s</unitLabel>
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/src/main/resources/OH-INF/config/config.xml:39]: To be suppressed.

File: bundles/org.openhab.voice.watsonstt/README.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+* **Speech Detector Sensitivity** - Use the parameter to suppress word insertions from music, coughing, and other non-speech events.
+* **Inactivity Timeout** - The time in seconds after which, if only silence (no speech) is detected in the audio, the connection is closed.
+* **Opt Out Logging** - By default, all IBM Watson™ services log requests and their results. Logging is done only to improve the services for future users. The logged data is not shared or made public.
+* **No Results Message** - Message to be told when no results.
+* **Smart Formatting** - If true, the service converts dates, times, series of digits and numbers, phone numbers, currency values, and internet addresses into more readable. (Not available for all locales)
+* **Redaction** - If true, the service redacts, or masks, numeric data from final transcripts. (Not available for all locales)
+
+### Configuration via a text file
+
+In case you would like to setup the service via a text file, create a new file in `$OPENHAB_ROOT/conf/services` named `googlestt.cfg`
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/README.md:37]: Propose another file name

File: bundles/org.openhab.voice.watsonstt/pom.xml [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.3.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>org.openhab.voice.watsonstt</artifactId>
+
+  <name>openHAB Add-ons :: Bundles :: WatsonSTT Binding</name>
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/pom.xml:15]: openHAB Add-ons :: Bundles :: Voice :: IBM Watson Speech to Text

File: bundles/org.openhab.voice.watsonstt/README.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+```
+org.openhab.voice.watsonstt:apiKey=******
+org.openhab.voice.watsonstt:instanceUrl=https://api.***.speech-to-text.watson.cloud.ibm.com/instances/*****
+org.openhab.voice.watsonstt:backgroundAudioSuppression=0.5
+org.openhab.voice.watsonstt:speechDetectorSensitivity=0.5
+org.openhab.voice.watsonstt:inactivityTimeout=2
+org.openhab.voice.watsonstt:optOutLogging=false
+org.openhab.voice.watsonstt:noResultsMessage="Sorry, I didn't understand you"
+```
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/README.md:49]: Please add an example for the two missing parameters.

File: bundles/org.openhab.voice.watsonstt/src/main/java/org/openhab/voice/watsonstt/internal/WatsonSTTConstants.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+ * used across the whole binding.
+ *
+ * @author Miguel Álvarez - Initial contribution
+ */
+@NonNullByDefault
+public class WatsonSTTConstants {
+    /**
+     * Service name
+     */
+    public static final String SERVICE_NAME = "IBM Watson Speech-to-Text";
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/src/main/java/org/openhab/voice/watsonstt/internal/WatsonSTTConstants.java:28]: Just "IBM Watson"

File: bundles/org.openhab.voice.watsonstt/src/main/java/org/openhab/voice/watsonstt/internal/WatsonSTTService.java [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+import okhttp3.WebSocket;
+
+/**
+ * The {@link WatsonSTTService} allows to use Watson as Speech-to-Text engine
+ *
+ * @author Miguel Álvarez - Initial contribution
+ */
+@NonNullByDefault
+@Component(configurationPid = SERVICE_PID, property = Constants.SERVICE_PID + "=" + SERVICE_PID)
+@ConfigurableService(category = SERVICE_CATEGORY, label = SERVICE_NAME, description_uri = SERVICE_CATEGORY + ":"
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/src/main/java/org/openhab/voice/watsonstt/internal/WatsonSTTService.java:66]: label = SERVICE_NAME + " Speech-to-Text"

File: bundles/org.openhab.voice.watsonstt/src/main/java/org/openhab/voice/watsonstt/internal/WatsonSTTService.java [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+import com.ibm.cloud.sdk.core.security.IamAuthenticator;
+import com.ibm.watson.speech_to_text.v1.SpeechToText;
+import com.ibm.watson.speech_to_text.v1.model.RecognizeWithWebsocketsOptions;
+import com.ibm.watson.speech_to_text.v1.model.SpeechRecognitionAlternative;
+import com.ibm.watson.speech_to_text.v1.model.SpeechRecognitionResult;
+import com.ibm.watson.speech_to_text.v1.model.SpeechRecognitionResults;
+import com.ibm.watson.speech_to_text.v1.websocket.RecognizeCallback;
+
+import okhttp3.OkHttpClient;
+import okhttp3.WebSocket;
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/src/main/java/org/openhab/voice/watsonstt/internal/WatsonSTTService.java:57]: Is it possible to not use okhttp3 ?
This is not the recommended HTTP client for openHAB.

File: bundles/org.openhab.voice.watsonstt/src/main/resources/OH-INF/config/config.xml [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+			<description>By default, all IBM Watson™ services log requests and their results. Logging is done only to improve the
+				services for future users. The logged data is not shared or made public.</description>
+			<default>true</default>
+		</parameter>
+		<parameter name="smartFormatting" type="boolean" groupName="stt">
+			<label>Smart Formatting</label>
+			<description>If true, the service converts dates, times, series of digits and numbers, phone numbers, currency
+				values, and internet addresses into more readable. (Not available for all locales)</description>
+			<default>false</default>
+			<advanced>true</advanced>
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/src/main/resources/OH-INF/config/config.xml:57]: I have a doubt if this tag is deprecated or not. I will have to search.

**Review by lolodomo — CHANGES_REQUESTED** [review]
Review part 1 of 2

**Comment by GiviMAD** [comment]
I have added a retry on ssl exception because it happens eventually, and always works after retrying.
I won't be able to work more on this till Wednesday.

File: bundles/org.openhab.voice.watsonstt/src/main/java/org/openhab/voice/watsonstt/internal/WatsonSTTService.java (hunk 13 above)
Comment (GiviMAD) [reply bundles/org.openhab.voice.watsonstt/src/main/java/org/openhab/voice/watsonstt/internal/WatsonSTTService.java:59]: Those are the libraries used by the SDK.
I think that rewrite this add-on using the preferred http and websocket clients shouldn't be hard. I will give it a try in the future.
But, as this is already working, I will prefer to keep this first version using the SDK. Is this ok for you, @lolodomo?

Comment (lolodomo) [reply bundles/org.openhab.voice.watsonstt/src/main/java/org/openhab/voice/watsonstt/internal/WatsonSTTService.java:59]: If this is required by the underlying DDK, I understand you have no choice.

File: bundles/org.openhab.voice.watsonstt/src/main/resources/OH-INF/config/config.xml (hunk 14 above)
Comment (lolodomo) [reply bundles/org.openhab.voice.watsonstt/src/main/resources/OH-INF/config/config.xml:57]: Checked. This is perfectly correct.

File: bundles/org.openhab.voice.watsonstt/src/main/resources/OH-INF/config/config.xml (hunk 7 above)
Comment (GiviMAD) [reply bundles/org.openhab.voice.watsonstt/src/main/resources/OH-INF/config/config.xml:39]: I don't understand the comment. Can you add some explanation please?

File: bundles/org.openhab.voice.watsonstt/README.md (hunk 4 above)
Comment (GiviMAD) [reply bundles/org.openhab.voice.watsonstt/README.md:11]: No, I'm using one and I can't see any notice about the service going down.

**Comment by GiviMAD** [comment]
@lolodomo I just realize this add-on doesn't work with the 3.3.0M1 due to https://github.com/openhab/openhab-core/issues/2702, as the library I'm using call the close of the audio stream. 
I can workaround it wrapping the stream in another one that does nothing on close, or is it better that I wait for the issue to be solved?
Thanks for the help.

**Comment by lolodomo** [comment]
It should not work better in any 3.x version. 
Yes, the problem has to be fixed first in the core framework
 If you have a solution, please submit a PR.

**Comment by GiviMAD** [comment]
> It should not work better in any 3.x version.
The concrete problem is that it works for a single execution as it close the shared mic which remains closed until I restart openHAB.
Just wanted to inform you in case you are already looking at it, I will open a pr to stop sharing the mic, hope it's accepted.
Again, thanks all the help on this matter.

File: bundles/org.openhab.voice.watsonstt/src/main/resources/OH-INF/config/config.xml (hunk 7 above)
Comment (lolodomo) [reply bundles/org.openhab.voice.watsonstt/src/main/resources/OH-INF/config/config.xml:39]: The unit is already defined as an attribute.
What means this tag for ?

File: bundles/org.openhab.voice.watsonstt/src/main/java/org/openhab/voice/watsonstt/internal/WatsonSTTService.java [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+            "en-GB_BroadbandModel", "en-US_BroadbandModel", "es-AR_BroadbandModel", "es-CL_BroadbandModel",
+            "es-CO_BroadbandModel", "es-ES_BroadbandModel", "es-MX_BroadbandModel", "es-PE_BroadbandModel",
+            "fr-CA_BroadbandModel", "fr-FR_BroadbandModel", "it-IT_BroadbandModel", "ja-JP_BroadbandModel",
+            "ko-KR_BroadbandModel", "nl-NL_BroadbandModel", "pt-BR_BroadbandModel", "zh-CN_BroadbandModel");
+    private final Set<Locale> supportedLocales = models.stream().map(name -> name.split("_")[0])
+            .map(Locale::forLanguageTag).collect(Collectors.toSet());
+    private WatsonSTTConfiguration config = new WatsonSTTConfiguration();
+
+    @Activate
+    protected void activate(Map<String, Object> config) {
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/src/main/java/org/openhab/voice/watsonstt/internal/WatsonSTTService.java:81]: Please also implement (override) the method `modified` which will be called when the service configuration is updated.

**Review by lolodomo — CHANGES_REQUESTED** [review]
Review part 2 of 2

**Comment by lolodomo** [comment]
Are you able to produce the default translations properties ?

File: bundles/org.openhab.voice.watsonstt/pom.xml [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+      <artifactId>okio</artifactId>
+      <version>2.8.0</version>
+      <scope>compile</scope>
+    </dependency>
+    <dependency>
+      <groupId>com.google.code.gson</groupId>
+      <artifactId>gson</artifactId>
+      <version>2.8.9</version>
+      <scope>compile</scope>
+    </dependency>
```
[the remark this answers is not in the document]
Comment (GiviMAD) [reply bundles/org.openhab.voice.watsonstt/pom.xml:68]: Just tested, I'm not able to make it work without it.

File: bundles/org.openhab.voice.watsonstt/src/main/resources/OH-INF/i18n/watsonstt.properties [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+voice.config.watsonstt.noResultsMessage.description = Message to be told when no transcription is done.
+voice.config.watsonstt.optOutLogging.label = Opt Out Logging
+voice.config.watsonstt.optOutLogging.description = By default, all IBM Watson™ services log requests and their results. Logging is done only to improve the services for future users. The logged data is not shared or made public.
+voice.config.watsonstt.redaction.label = Redaction
+voice.config.watsonstt.redaction.description = If true, the service redacts, or masks, numeric data from final transcripts. (Not available for all locales)
+voice.config.watsonstt.smartFormatting.label = Smart Formatting
+voice.config.watsonstt.smartFormatting.description = If true, the service converts dates, times, series of digits and numbers, phone numbers, currency values, and internet addresses into more readable. (Not available for all locales)
+voice.config.watsonstt.speechDetectorSensitivity.label = Speech Detector Sensitivity
+voice.config.watsonstt.speechDetectorSensitivity.description = Use the parameter to suppress word insertions from music, coughing, and other non-speech events.
```
Comment (lolodomo) [inline bundles/org.openhab.voice.watsonstt/src/main/resources/OH-INF/i18n/watsonstt.properties:22]: Please add these lines at the end of this file:
```

# service

service.voice.watsonstt.label = IBM Watson Speech-to-Text
```

**Review by lolodomo — CHANGES_REQUESTED** [review]
Last request

**Comment by lolodomo** [comment]
A return is missing at the end of the properties file.

**Comment by GiviMAD** [comment]
Done, thanks for warn me.

**Comment by lolodomo** [comment]
@GiviMAD : thank you for your new contribution/

**Pull request merged**